### PR TITLE
Design

### PR DESCRIPTION
**Pull request by mitchelljkotler**

### CODE REVIEW

File: squarelet/organizations/views.py [hunk 1]
```diff
@@ -173,31 +233,90 @@ class ManageMembers(OrganizationAdminMixin, UpdateView):
...
+                    organization=organization, email=addmember_form.cleaned_data["email"]
+                )
+                invitation.send()
+                messages.success(self.request, "Invitation sent")
+        return redirect('organizations:manage-members', slug=self.get_object().slug)
+
+    def _handle_invite(self, request, inviteFn, successMessage):
+        try:
+            inviteid = request.POST.get('inviteid')
+            assert(inviteid is not None)
```
Comment (mitchelljkotler) [inline squarelet/organizations/views.py]: `assert`s should not be used for validation

File: squarelet/organizations/views.py [hunk 2]
```diff
@@ -173,31 +233,90 @@ class ManageMembers(OrganizationAdminMixin, UpdateView):
...
+        elif action == 'removeuser':
+            return self._handle_user(request, lambda membership: membership.delete(), 'Removed user')
+
+    def _handle_add_member(self, request, addmember_form):
+        if not addmember_form.is_valid():
+            # Ensure the email is valid
+            messages.error(request, "Please enter a valid email address")
+        else:
+            # Ensure the org has capacity
+            organization = self.get_object()
```
Comment (mitchelljkotler) [inline squarelet/organizations/views.py]: `get_object` is slow, call once at beginning of `post`

File: squarelet/organizations/views.py [hunk 3]
```diff
@@ -173,31 +233,90 @@ class ManageMembers(OrganizationAdminMixin, UpdateView):
...
+                invitation.send()
+                messages.success(self.request, "Invitation sent")
+        return redirect('organizations:manage-members', slug=self.get_object().slug)
+
+    def _handle_invite(self, request, inviteFn, successMessage):
+        try:
+            inviteid = request.POST.get('inviteid')
+            assert(inviteid is not None)
+            invite = Invitation.objects.get(pk=inviteid)
+            assert(invite.organization == self.get_object())
```
Comment (mitchelljkotler) [inline squarelet/organizations/views.py]: You need to make sure the Inivitation exists - they could pass a bad ID

File: squarelet/organizations/views.py [hunk 4]
```diff
@@ -173,31 +233,90 @@ class ManageMembers(OrganizationAdminMixin, UpdateView):
...
     def get_context_data(self, **kwargs):
         context = super().get_context_data(**kwargs)
         form = context["form"]
-        context["users"] = [
-            (u, form[f"admin-{u.pk}"], form[f"remove-{u.pk}"])
-            for u in self.object.users.all()
-        ]
+        context["admin"] = self.request.user
+        users = self.object.users.all().order_by('created_at')
+        context["users"] = [(u, self.get_object().memberships.get(user_id=u.id).admin) for u in users]
```
Comment (mitchelljkotler) [inline squarelet/organizations/views.py]: This causes N+1 DB queries (one query for each user)

This does it in one query:
```
         context["users"] = [
             (m.user, m.admin)
             for m in self.organization.memberships.select_related("user").order_by(
                 "user__created_at"
             )
         ]
```

File: squarelet/organizations/views.py [hunk 5]
```diff
@@ -173,31 +233,90 @@ class ManageMembers(OrganizationAdminMixin, UpdateView):
...
+        context["requested_invitations"] = self.object.invitations.get_requested().order_by('created_at')
```
Comment (mitchelljkotler) [inline squarelet/organizations/views.py]: Probably makes sense to just make `created_at` the default ordering for invitations

File: squarelet/templates/account/field.html [hunk 6]
```diff
@@ -0,0 +1,11 @@
+{% load crispy_forms_field %}
+
+{% if field.is_hidden %}
+  {{ field }}
+{% else %}
+  <div class="_cls-field">
+    {% crispy_field field %}
+    <!-- <input type="{{field.input_type}}" class="{{field.field_class}}" placeholder="{{field.label}}"> -->
```
Comment (mitchelljkotler) [inline squarelet/templates/account/field.html]: remove comment?

File: squarelet/static/js/project.js [hunk 7]
```diff
@@ -18,4 +18,4 @@ Issues with the above approach:
 3. Really Unforgiving: No way to override this behavior
 4. Undocumented: No mention in the documentation, or it's too hard for me to find
 */
-$('.form-group').removeClass('row');
+// $('.form-group').removeClass('row');
```
Comment (mitchelljkotler) [inline squarelet/static/js/project.js]: just remove this file?

File: squarelet/templates/account/login.html [hunk 8]
```diff
@@ -3,18 +3,45 @@
...
 
+<div class="_cls-content">
+  {% trans "Sign in for" as intent_header %}
+  {% trans "You’ll also be able to log into" as intent_text %}
+  {% handleintent intent_header intent_text %}
+
+  <!-- Start form -->
+  <form id="login_form" method="POST" action="{% url 'account_login' %}">
+    {% crispy form form.helper %}
+    {% comment %} <div class="_cls-field">
```
Comment (mitchelljkotler) [inline squarelet/templates/account/login.html]: remove commented code?

File: squarelet/templates/account/login.html [hunk 9]
```diff
@@ -3,18 +3,45 @@
...
+    <div class="_cls-actionBig">
+      <button>{% trans 'Log in' %}</button>
+    </div>
+    <div class="_cls-center">
+      <div class="_cls-minorInfo">{% blocktrans %}Don’t have an account?{% endblocktrans %}</div>
+      <a href="{% url 'select_plan' %}"><div class="_cls-action">{% blocktrans %}Sign&nbsp;Up{% endblocktrans %}</div></a>
+    </div>
+  </form>
+</div>
+{% comment %}
```
Comment (mitchelljkotler) [inline squarelet/templates/account/login.html]: remove commented code?

File: squarelet/templates/base.html [hunk 10]
```diff
@@ -99,9 +99,9 @@
       {% block javascript %}
 
         <!-- Required by Bootstrap v4 beta -->
-        <script src="https://code.jquery.com/jquery-3.2.1.slim.min.js" integrity="sha384-KJ3o2DKtIkvYIK3UENzmM7KCkRr/rE9/Qpg6aAZGJwFDMVNA/GpGFF93hXpG5KkN" crossorigin="anonymous"></script>
+        <!-- <script src="https://code.jquery.com/jquery-3.2.1.slim.min.js" integrity="sha384-KJ3o2DKtIkvYIK3UENzmM7KCkRr/rE9/Qpg6aAZGJwFDMVNA/GpGFF93hXpG5KkN" crossorigin="anonymous"></script>
```
Comment (mitchelljkotler) [inline squarelet/templates/base.html]: remove?

File: squarelet/templates/base.html [hunk 11]
```diff
@@ -99,9 +99,9 @@
...
         <script src="https://cdnjs.cloudflare.com/ajax/libs/popper.js/1.11.0/umd/popper.min.js" integrity="sha384-b/U6ypiBEHpOf/4+1nzFpr53nxSS+GLCkfwBdFNTxtclqqenISfwAzpKaMNFNmj4" crossorigin="anonymous"></script>
-        <script src="https://stackpath.bootstrapcdn.com/bootstrap/4.1.3/js/bootstrap.min.js" integrity="sha384-ChfqqxuZUCnJSK3+MXmPNIyE6ZbWh2IMqE241rYiqJxyMiZ6OW/JmZQ5stwEULTy" crossorigin="anonymous"></script>
+        <script src="https://stackpath.bootstrapcdn.com/bootstrap/4.1.3/js/bootstrap.min.js" integrity="sha384-ChfqqxuZUCnJSK3+MXmPNIyE6ZbWh2IMqE241rYiqJxyMiZ6OW/JmZQ5stwEULTy" crossorigin="anonymous"></script> -->
```
Comment (mitchelljkotler) [inline squarelet/templates/base.html]: replace cdn js with a js build system?

File: squarelet/templates/organizations/organization_managemembers.html [hunk 12]
```diff
@@ -1,31 +1,157 @@
...
+    <form class="_cls-inlineForm" method="post" enctype="multipart/form-data" id="org-form">
+      {% csrf_token %}
+      <input type="hidden" name="action" value="managemembers">
+      <div class="_cls-manageTable">
+        <div class="_cls-manageRow _cls-manageHeader">
+          <div class="_cls-manageCell">{% trans 'User' %}</div>
+          <div class="_cls-manageCell">{% trans 'Admin' %}</div>
+          <div class="_cls-manageCell">{% trans 'Action' %}</div>
+          </div>
+        {% for user, isAdmin in users %}
```
Comment (mitchelljkotler) [inline squarelet/templates/organizations/organization_managemembers.html]: s/isAdmin/is_admin/g

File: squarelet/templates/users/user_detail.html [hunk 13]
```diff
@@ -1,44 +1,123 @@
...
-      {% empty %}
-        <p>No avatar</p>
-      {% endthumbnail %}
+  <div class="_cls-largeContent">
+    <div class="_cls-profileColumns">
+      <div class="_cls-profileColumn">
+        <div class="_cls-profileSection">
+          <div class="_cls-profileInfo">
+            <div class="_cls-profileAvatar">
+              {% autoescape off %}{% avatar user 37 %}{% endautoescape %}
```
Comment (mitchelljkotler) [inline squarelet/templates/users/user_detail.html]: you should not turn off autoescape here, build a safe string the template tag:

https://docs.djangoproject.com/en/2.1/ref/utils/#django.utils.html.format_html

File: squarelet/core/templatetags/bold.py [hunk 14]
```diff
@@ -0,0 +1,28 @@
...
+
+# Standard Library
+import html
+
+register = template.Library()
+
+
+@register.simple_tag
+def bold(highlight_text, full_text):
+    text = html.escape(full_text)
```
Comment (mitchelljkotler) [inline squarelet/core/templatetags/bold.py]: Use https://docs.djangoproject.com/en/2.1/ref/utils/#django.utils.html.format_html

File: squarelet/core/templatetags/bold.py [hunk 15]
```diff
@@ -0,0 +1,28 @@
...
+    if highlight_text == "":
+        return text
+
+    result = ""
+
+    while True:
```
[the remark this answers is not in the document]
Comment (mitchelljkotler) [reply squarelet/core/templatetags/bold.py]: On second look, do you even use this anywhere?  I can't find it, and getting the escaping right is tricky - if not used it should be removed

File: squarelet/templates/base.html (hunk 11 above)
Comment (freedmand) [reply squarelet/templates/base.html]: The CDN JS is no longer used. It's pretty sparse in the JS land (I think just autocomplete.js), but I need to figure out the build system. I'm leaning using Rollup for this. I can publish my minimal repo for the autocomplete code or just integrate it in a JS folder

**Comment by freedmand** [comment]
Thanks for this review -- it's really helpful. I have a lot to learn in Django-land but feel a lot more confident going forward. Will tackle some of these issues more in the morning

**Pull request merged**